src/go/util/httppattern/match_sequence_generator.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httppattern
+
+import (
+	"fmt"
+	"strings"
+)
+
+// MatchSequenceGenerator generate the match sequence of added methods based on their
+// http pattern. The sequence should be same as the matching order of
+// https://github.com/GoogleCloudPlatform/esp-v2/blob/641ce1d5c177401e424f2b27dd45de1bf797530b/src/api_proxy/path_matcher/path_matcher.h#L1
+type MatchSequenceGenerator struct {
+	RootPtr     *matchSequenceGeneratorNode
+	CustomVerbs map[string]bool
+}
+
+type Method struct {
+	UriTemplate string
+	HttpMethod  string
+	Operation   string
+}
+
+func NewMatchSequenceGenerator() *MatchSequenceGenerator {
+	return &MatchSequenceGenerator{
+		RootPtr:     newMatchSequenceGeneratorNode(),
+		CustomVerbs: make(map[string]bool),
+	}
+}
+
+func (h *MatchSequenceGenerator) Register(method *Method) error {
+	if method == nil {
+		return fmt.Errorf("empty method")
+	}
+
+	uriTemplate := method.UriTemplate
+	httpMethod := method.HttpMethod
+
+	ht := Parse(uriTemplate)
+	if ht == nil {
+		return fmt.Errorf("invalid url template `%s`", uriTemplate)
+	}
+
+	pathInfo := transferFromUriTemplate(ht)
+	methodData := &methodData{
+		Method:   method,
+		Variable: ht.Variables,
+	}
+
+	if !h.RootPtr.insertPath(pathInfo, httpMethod, methodData, true) {
+		return fmt.Errorf("duplicate http pattern `%s %s`", httpMethod, uriTemplate)
+	}
+
+	if ht.Verb != "" {
+		h.CustomVerbs[ht.Verb] = true
+	}
+
+	return nil
+}
+
+type MatchSequence []*Method
```

Comment:
Nit: Prefer if all public types are defined at the top of the file. It's easy to miss this.